crates/ruma-identifiers/src/macros.rs

```diff
@@ -24,8 +24,63 @@ macro_rules! partial_eq_string {
     }
 }
 
+macro_rules! owned_identifier {
+    ($owned:ident, $id:ident) => {
+        #[doc = concat!("Owned variant of ", stringify!($id))]
+        #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+        pub struct $owned {
+            #[cfg(not(ruma_identifiers_storage))]
+            inner: Box<$id>,
+            #[cfg(ruma_identifiers_storage = "Arc")]
+            inner: std::sync::Arc<$id>,
+        }
+
+        impl AsRef<$id> for $owned {
+            fn as_ref(&self) -> &$id {
+                &self.inner
+            }
+        }
+
+        impl std::borrow::Borrow<$id> for $owned {
+            fn borrow(&self) -> &$id {
+                self.as_ref()
+            }
+        }
+
+        impl From<&'_ $id> for $owned {
+            fn from(id: &$id) -> $owned {
+                let boxed: Box<$id> = id.into();
+
+                boxed.into()
```

Comment:
Would this work instead? Would do one allocation instead of two when inner is `Arc`:

```suggestion
                $owned { inner: id.into() }
```

Reply:
Maybe for `(*id).into()`, as there is no `From<&T>` for box or arc.

I'll try to see if that works